Introduce multi-Kueue support for ElasticJobs, with specific support for `batch/v1.Job` workloads when ElasticJob functionality is enabled.

ichekrygin opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

Now that the core ElasticJobsViaWorkloadSlices functionality has been introduced in Kueue as part of [KEP-77](https://github.com/kubernetes-sigs/kueue/tree/main/keps/77-dynamically-sized-jobs), this PR introduces MultiKueue support for ElasticJobsViaWorkloadSlices.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #6335 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Add: Multikueue support for ElasticJobsViaWorklaodSlices with `batchv1/Job` support.
```

ichekrygin commented:
/retest

mimowo commented on pkg/workloadslicing/workloadslicing.go:
```diff
@@ -66,26 +66,34 @@ func Enabled(object metav1.Object) bool {
+// was scheduled in a multikueue context. If the corresponding annotation is not set,
+// it returns an empty string.
+func ClusterName(wl *kueue.Workload) string {
+	return wl.GetAnnotations()[WorkloadSliceClusterName]
```
Why not use `workload.status.clusterName`? This field was introduced in 0.13

ichekrygin replied:
> I would argue against extending API surface until other options are exhausted.

:100:

> Creating annotation as an existing field will raise huge confusion I think among users and when debugging. So please investigate deeper the option of reusing what we have.

Agreed. Perhaps all we need is to do due diligence on current usage and update the API docs to clarify the intent of the field. That might be sufficient to align expectations without introducing new fields or annotations.

ichekrygin replied:
Updated. PTAL: specific focus on type documentation.

GonzaloSaez commented on pkg/controller/jobs/job/job_multikueue_adapter.go:
```diff
@@ -38,12 +38,33 @@ import (
+	}
+	value, found := labels[labelKey]
+	labels[labelKey] = labelValue
+	return found && value == labelValue
```
I think this should be `!found || value != labelValue`

tenzen-y commented on pkg/workloadslicing/workloadslicing.go:
```diff
@@ -66,26 +66,34 @@ func Enabled(object metav1.Object) bool {
 	return &ref
 }
 
+// ClusterName returns the name of the remote cluster where the original workload
+// was scheduled in a multikueue context. If the corresponding annotation is not set,
+// it returns an empty string.
+func ClusterName(wl *kueue.Workload) string {
+	return wl.GetAnnotations()[WorkloadSliceClusterName]
```
If we need to add a different cluster nomination mechanism from non-WorkloadSlice usage, I would request updating KEP-77.

ichekrygin commented:
/retest

ichekrygin commented:
Hello @mimowo, @tenzen-y, @GonzaloSaez - I updated the PR to address comments, PTAL (don't want this PR to go stale) :)

tenzen-y commented on pkg/controller/jobs/job/job_multikueue_adapter.go:
```diff
@@ -38,12 +38,33 @@ import (
 
 var _ jobframework.MultiKueueAdapter = (*multiKueueAdapter)(nil)
 
+// updateLabel helper sets the provided label key/value on the provided job.
+// Returns true, if the job's label was changed.
+func updateLabel(job *batchv1.Job, labelKey, labelValue string) bool {
+	labels := job.Labels
+	if labels == nil {
+		job.Labels = map[string]string{labelKey: labelValue}
+		return true
+	}
+	value, found := labels[labelKey]
+	labels[labelKey] = labelValue
+	return !found || value != labelValue
+}
+
```
```suggestion
```

I think that we can rely on `utilmaps.Copy()` function. See the following comment.

tenzen-y commented on pkg/controller/jobs/job/job_multikueue_adapter.go:
```diff
@@ -38,12 +38,33 @@ import (
+	return !found || value != labelValue
+}
+
+// updateParallelism helper sets the provided parallelism on the provided Job (spec).
+// Returns true, if the job's parallelism value was changed.
+func updateParallelism(job *batchv1.Job, parallelism *int32) bool {
+	job.Spec.Parallelism, parallelism = parallelism, job.Spec.Parallelism
+	return ptr.Deref(job.Spec.Parallelism, 0) != ptr.Deref(parallelism, 0)
+}
```
This is called once, then this function does not seem to have much responsibility.
So, I would postpone having such util function.
I proposed how to update this in the following comment.

ichekrygin replied:
Answered in https://github.com/kubernetes-sigs/kueue/pull/6445#discussion_r2302640674

tenzen-y commented on pkg/controller/jobs/job/job_multikueue_adapter.go:
```diff
@@ -85,6 +106,17 @@ func (b *multiKueueAdapter) SyncJob(ctx context.Context, localClient client.Clie
+		if workloadslicing.Enabled(&localJob) {
+			// Update remote job's workload slice name and parallelism if needed.
+			if err := clientutil.Patch(ctx, remoteClient, &remoteJob, true, func() (bool, error) {
+				updatedLabels := updateLabel(&remoteJob, constants.PrebuiltWorkloadLabel, workloadName)
+				return updateParallelism(&remoteJob, localJob.Spec.Parallelism) || updatedLabels, nil
```
```suggestion
				originalLabels := maps.Clone(remoteJob.Labels)
				utilmaps.Copy(&remoteJob.Labels, map[string]string{constants.PrebuiltWorkloadLabel: workloadName})
				originalParallelism := cmp.Or(localJob.Spec.Parallelism, ptr.To[int32](0))
				remoteJob.Spec.Parallelism = localJob.Spec.Parallelism
				return !ptr.Equal(&originalParallelism, remotejob.Spec.Parallelism) || !maps.Equal(originalLabels, remoteJob.Labels), nil
```

This is for w/o `func updateLabel()` and `func updateParallelism()`.

tenzen-y commented on apis/kueue/v1beta1/workload_types.go:
```diff
@@ -424,7 +424,12 @@ type WorkloadStatus struct {
 	NominatedClusterNames []string `json:"nominatedClusterNames,omitempty"`
 
-	// clusterName is the name of the cluster where the workload is actually assigned.
+	// clusterName is the name of the cluster where the workload is currently assigned.
+	//
+	// With ElasticJobs, this field may also indicate the cluster where the original (old) workload
+	// was assigned, providing placement context for new scaled-up workloads. This supports
+	// affinity or propagation policies across workload slices.
+	//
 	// This field is reset after the Workload is evicted.
 	// +optional
 	ClusterName *string `json:"clusterName,omitempty"`
```
After I checked the implementations, I understand what happens when combining MultiKueue and DynamicJob.
Could you check ?

ichekrygin replied:
I am resolving this to continue discussion in https://github.com/kubernetes-sigs/kueue/pull/6445#discussion_r2301727074

tenzen-y commented on pkg/controller/jobs/job/job_multikueue_adapter.go:
Could you update UTs for TestMultiKueueAdapter in pkg/controller/jobs/job/job_multikueue_adapter_test.go?

ichekrygin replied:
Updated, PTAL.
Note: instead of modifying the existing `TestMultiKueueAdapter`, I created a new `Test_multiKueueAdapter_SyncJob` that follows a more Go-idiomatic testing style and improves coverage. The new test reaches "100%" coverage—matching the original test while adding coverage for ElasticJobs. If we like this approach, we can replace the old test with it. If not, I can trim down the duplicated coverage and limit the new test to just the missing ElasticJobs cases.

tenzen-y commented on pkg/controller/jobframework/validation.go:
Can you update UTs in pkg/controller/jobs/job/job_webhook_test.go?

ichekrygin commented on pkg/controller/admissionchecks/multikueue/workload.go:
```diff
@@ -451,45 +486,53 @@ func (w *wlReconciler) nominateAndSynchronizeWorkers(ctx context.Context, group
+	// In the future, we may introduce more nuanced remote workload propagation policies,
+	// supporting preferred or required placement constraints.
+	if clusterName := workloadslicing.ClusterName(group.local); group.IsElasticWorkload() && clusterName != "" {
+		nominatedWorkers = []string{clusterName}
```
> Can you store the enforced cluster name in the workload status?

Not entirely sure if I understand the question. in this code, we don't persist cluster name into status value (in K8s API/etc). We simply using it as a "short-circuit" nomination. 

* **Non-elastic workloads**, we follow the standard nomination flow where each remote cluster can be added to the nominated set.
* **Elastic workloads**, new slice scale-up, the slice “inherits” the original workload’s cluster and targets it directly, effectively bypassing multi-cluster nomination for that slice.

ichekrygin replied:
This comment is closely relate to https://github.com/kubernetes-sigs/kueue/pull/6445#discussion_r2301727074 so let's continue conversation there.

ichekrygin commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -827,12 +807,39 @@ func (r *JobReconciler) ensureOneWorkload(ctx context.Context, job GenericJob, o
 			return nil, err
 		}
 
+		// Skip workload-in-sync check for workload-slice since this could be a
+		// scaled-up (new) workload slice for the not-yet synced job object.
+		if workloadslicing.Enabled(object) {
```
> Could you elaborate on that?

Absolutely. First, the context. We are talking about a **remote Workload** propagated by MultiKueue into a worker cluster, with a prebuilt name derived from the manager cluster.

In this flow we assert that the remote Workload stays in sync with the corresponding remote Job. For **regular (non-elastic) workloads**, any Job ⇄ Workload spec drift is not tolerated, the Workload is marked **Finished** and cleaned up.

**Elastic workloads** are different. A temporary mismatch in **replica count** between the “old” slice and the “new” slice is expected and tolerated until the new slice is admitted. This stems from a two-phase propagation model:

1. propagate the Workload,
2. on admission, sync Job changes.

Until admission happens, an elastic remote Workload can be briefly out of sync with the remote Job by design.

Concretely, if someone edits the **remote Job** replicas directly:

* **Increase replicas**: extra Pods may be created in a scheduling-gated state, they will be removed on the next MultiKueue sync.
* **Decrease replicas**: Pods may be deleted, the missing Pods will be recreated on the next MultiKueue sync.

Only the replica count drift for elastic transitions is tolerated; other spec divergences remain non-tolerated and will trigger the normal “mark Finished and prune” behavior.

ichekrygin replied:
This `if` section is removed. Resolving.

ichekrygin commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1203,11 +1210,17 @@ func ConstructWorkload(ctx context.Context, c client.Client, job GenericJob, lab
 
 // prepareWorkloadSlice adds necessary workload slice annotations.
 func prepareWorkloadSlice(ctx context.Context, clnt client.Client, job GenericJob, wl *kueue.Workload) error {
+	// Mark the workload with the elastic-job enabled annotation.
+	// This allows distinguishing whether elastic-job support is enabled
+	// for a given workload at the workload level.
+	metav1.SetMetaDataAnnotation(&wl.ObjectMeta, workloadslicing.EnabledAnnotationKey, workloadslicing.EnabledAnnotationValue)
```
This is a great question.

First, in the initial implementation of adding the support for **ElasticJobsViaWorkloadSlices**: `kueue.x-k8s.io/elastic-job: "true"` annotation is added to the **Job** object, not to the **Workload**.

In a single Kueue setup, a Workload is always a realization of a Job, that is, first there is a Job, then Kueue creates the corresponding Workload. Therefore we can determine whether a Workload is elastic by inspecting its parent Job.

In a MultiKueue setup on a remote worker cluster, the order is reversed, first there is a propagated Workload, then the Job is created upon admission. Since the Job may not exist yet, we cannot tell whether the Workload is elastic by looking for the Job. To address this, we copy the same “elastic job” annotation `kueue.x-k8s.io/elastic-job: "true"` onto every Workload that corresponds to an elastic Job, in both single Kueue and MultiKueue contexts, so elasticity is discoverable even before the Job materializes.

ichekrygin replied:
In the **MultiKueue** context, specifically in the MultiKueue workload reconciler, we don’t have immediate access to the Job object associated with a workload (unlike in the `GenericJobReconciler`). Still, we need to determine whether a workload corresponds to an **ElasticJob** in order to handle it correctly, for example whenever we invoke [this logic](https://github.com/kubernetes-sigs/kueue/blob/65c3d3f897004752c24910dcbe8435f344ab316b/pkg/controller/admissionchecks/multikueue/workload.go#L100)

As an optimization, this annotation is propagated to the workload at creation time.

mimowo replied:
> In the MultiKueue context, specifically in the MultiKueue workload reconciler, we don’t have immediate access to the Job object associated with a workload (unlike in the GenericJobReconciler). Still, we need to determine whether a workload corresponds to an ElasticJob in order to handle it correctly, for example whenever we invoke [this logic](https://github.com/kubernetes-sigs/kueue/blob/65c3d3f897004752c24910dcbe8435f344ab316b/pkg/controller/admissionchecks/multikueue/workload.go#L100)

Thank you for explaining!

> @mimowo Any thoughts?

I think `AnnotationKeysToCopy` looks like overkill. Also, users of MultiKueue expect it to work OOTB, they don't want to need to configure another place. `LabelKeysToCopy` is a user-facing mechanism for custom user labels, not for inner workings of Kueue. I think this annotation should be supported OOTB.

> So, you would like to embed the ElasticJob annotation to WorkloadSlice, right?

I think to reduce confusion we could consider naming it differently on the prebuilt-workload object, like:
`kueue.x-k8s.io/prebuilt-workload-for-elastic-job`. Wdyt?

ichekrygin commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -1224,6 +1237,9 @@ func prepareWorkloadSlice(ctx context.Context, clnt client.Client, job GenericJo
 		oldSlice := workloadSlices[0]
 		// Annotate new workload slice with the preemptible (old) workload slice.
 		metav1.SetMetaDataAnnotation(&wl.ObjectMeta, workloadslicing.WorkloadSliceReplacementFor, string(workload.Key(&oldSlice)))
+
+		// Keep the same cluster assignment between slices.
+		wl.Status.ClusterName = oldSlice.Status.ClusterName
```
**Cluster pinning on scale up**

During scale up, we copy the cluster name from the original Workload to the new Workload slice.

* **Single-queue setup**, `wl.Status.ClusterName` remains an empty string.
* **Multi-queue setup**, every subsequent Workload slice preserves the same cluster as the original slice, ensuring all slices of the Job target the same worker cluster.

This keeps scale-up slices co-located with the original workload and avoids cross-cluster re-nomination. If you ever observe a non-empty `Status.ClusterName` in a single-queue setup, that should be treated as a bug or stale state.

mimowo replied:
TBH, I like to use clusterName for pinning, going via the nomination phase seems unnecessary if we already know we want a specific cluster. It seems to require extra code in the dispatchers which would not be necessary otherwise (dispatchers are skipped). This will also generate adoption frition, because developers of external dispatchers would need to think they need to do "skip" for ElasticJobs.

tenzen-y replied:
The `status` field is the actual state, and `spec` is the desired state.
However, if we assign any clusterName to the status field w/o QuotaReservation, the .status.clusterName represents the desired state since the `.status.clusterName` is no longer actual cluster name reserved by the cluster. 

So, I think we can (1 leverage NCN in the dispatcher as I mentioned above, or (2 worker clusters ' kueue scheduler assigns their cluster name to WorkloadSlice by themselves while QuotaReservation. In that case, we can keep the `.status.clusterName` as the actual state.

However, (2 indicates bringing cluster assignment responsibilities to the Kueue scheduler, which is not good since technically, the Kueue-scheduler is responsible for everything, but we decouple responsibilities across multiple internal components.

ichekrygin commented on pkg/controller/admissionchecks/multikueue/workload.go:
```diff
@@ -345,6 +358,19 @@ func (w *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) (reco
+			}
+		}
+		return reconcile.Result{}, errors.Join(errs...)
+	}
```
Updated.

ichekrygin commented:
/retest

mimowo commented on pkg/controller/admissionchecks/multikueue/workload.go:
```diff
@@ -430,20 +465,28 @@ func (w *wlReconciler) nominateAndSynchronizeWorkers(ctx context.Context, group
+			for workerName := range group.remotes {
+				nominatedWorkers = append(nominatedWorkers, workerName)
+			}
+			if group.local.Status.ClusterName == nil && !equality.Semantic.DeepEqual(group.local.Status.NominatedClusterNames, nominatedWorkers) {
+				group.local.Status.NominatedClusterNames = nominatedWorkers
+				if err := workload.ApplyAdmissionStatus(ctx, w.client, group.local, true, w.clock); err != nil {
+					log.V(2).Error(err, "Failed to patch nominated clusters", "workload", klog.KObj(group.local))
+					return reconcile.Result{}, err
+				}
```
This looks racy, as we update the NominatedClusterNames for the replacement workload which is already created. So, before we act it may get assigned to another cluster.

I think the proper solution would be to set the clusterName up front when creating the replacement Workload. Then, basically skip the nomination phase. Wdyt?

ichekrygin replied:
wonder if I messed it up during rebase. 
I will take another (closer) look.

ichekrygin replied:
I am sorry, responded incorrectly about "diff" here, instead of next comment. I moved my message(s) to[ the correct thread.](https://github.com/kubernetes-sigs/kueue/pull/6445#pullrequestreview-3241375600)

mimowo commented on pkg/controller/admissionchecks/multikueue/workload.go:
```diff
@@ -430,20 +465,28 @@ func (w *wlReconciler) nominateAndSynchronizeWorkers(ctx context.Context, group
+	if clusterName := workload.ClusterName(group.local); group.IsElasticWorkload() && clusterName != "" {
+		nominatedWorkers = []string{clusterName}
+	} else {
+		if w.dispatcherName == config.MultiKueueDispatcherModeAllAtOnce {
```
Related comment, but I don't like the dispatcher code needs to be changed. As mentioned in https://github.com/kubernetes-sigs/kueue/pull/6445/files#r2360609035 this will generate firction to require developers of external dispatchers to do some custom code. In case of ElasticJobs, at least for MVP, we assume pinning to the same cluster. So, I see no point going via the nomination phase, and complicating the code of disaptchers. cc @tenzen-y

ichekrygin replied:
Let's continue in https://github.com/kubernetes-sigs/kueue/pull/6445#discussion_r2360589742 to avoid duplication in comment.

mimowo commented:
Ok, I have given the PR the first pass. Overall LGTM, but I have two major open "design" questions (also asked in comments):
1. why the dispatcher's code is modified? Will developers of other dispatchers need to do the same? I'm sceptical of design which carries this overhead. I would like to skip the nomination phase, and rely directly on the clusterName.
2. why we need to copy the Job's annotation indicating this is elastic Job to Workload. This seems to make the semantic very tricky. I will need to understand *why*. If we really really need it, then fine, but I would like to consider another name to clearly state the difference.

ichekrygin commented on pkg/controller/admissionchecks/multikueue/workload.go:
```diff
@@ -313,8 +321,25 @@ func (w *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) (reco
 
-	// 1. delete all remote workloads when finished or the local wl has no reservation
-	if group.IsFinished() || !workload.HasQuotaReservation(group.local) {
+	// 1.1 Handle finished local workload by removing remote workloads and jobs.
+	if group.IsFinished() {
+		// Skip replaced elastic workloads finished workload slices.
+		if group.IsElasticWorkload() && workloadslicing.IsReplaced(group.local.Status) {
+			return reconcile.Result{}, nil
+		}
+		var errs []error
+		for rem := range group.remotes {
+			if err := group.RemoveRemoteObjects(ctx, rem); err != nil {
+				errs = append(errs, err)
+				log.V(2).Error(err, "Deleting remote workload", "workerCluster", rem)
+			}
+		}
+		return reconcile.Result{}, errors.Join(errs...)
+	}
```
Let’s step back and review the original logic:

```go
// 1. delete all remote workloads when finished or the local wl has no reservation
if group.IsFinished() || !workload.HasQuotaReservation(group.local) {
	var errs []error
	for rem := range group.remotes {
		if err := group.RemoveRemoteObjects(ctx, rem); err != nil {
			errs = append(errs, err)
			log.V(2).Error(err, "Deleting remote workload", "workerCluster", rem)
		}
	}
	return reconcile.Result{}, errors.Join(errs...)
}
```

In this code, we remove remote workloads when **either** of the following conditions is met:

* The local workload is finished, **OR**
* The local workload does **not** hold a quota reservation.

However, **ElasticWorkloads** require slightly more nuanced handling for "IsFinished" condition:

### Finished Workloads

For ElasticWorkloads, we remove remote workloads **only if** the local workload has finished "naturally", meaning via **completion, cancellation, or preemption**, etc.

If the workload is finished due to a **Workload[slice] Replacement** (as part of a scale-up event), we do **not** remove the remote objects. Instead, we ignore the "old" (i.e., replaced) workload, and simply return:

```go
return reconcile.Result{}, nil
```

No further cleanup of remotes is needed in this case.

### Workloads Without Quota Reservation

This path remains unchanged: if the local workload no longer holds a quota reservation, we remove all remote workloads, same as the original logic.

### Alternative

There’s no clean way to fold this logic into a single `if` clause without reducing clarity. So as an alternative, we could explicitly gate the "replacement-finished" case upfront:

```go
// 0. Ignore Elastic workloads finished due to Replacement.
if group.IsFinished() && workloadslicing.IsReplaced(group.local.Status) {
	return reconcile.Result{}, nil
}

// 1. Delete all remote workloads when finished or quota reservation is missing.
if group.IsFinished() || !workload.HasQuotaReservation(group.local) {
	var errs []error
	for rem := range group.remotes {
		if err := group.RemoveRemoteObjects(ctx, rem); err != nil {
			errs = append(errs, err)
			log.V(2).Error(err, "Deleting remote workload", "workerCluster", rem)
		}
	}
	return reconcile.Result{}, errors.Join(errs...)
}
```

I'm happy to make this change if we find it more acceptable.

ichekrygin commented on pkg/controller/admissionchecks/multikueue/workload.go:
done.

ichekrygin commented:
@mimowo, @tenzen-y, thank you for the review and thoughtful comments.

I addressed all questions and comments, some of which resulted in code changes. Summary:

* **Dispatcher code appears modified**

  * Clarified that the dispatcher logic is unchanged, it is now nested under a new `if/else`, so diffs may look noisy unless whitespace changes are ignored.

* **Why copy the ElasticJob annotation from Job to Workload**

  * Explained the need to read this signal in the MultiKueue workload reconciler without fetching the Job each time, and noted the alternative of on demand Job lookups.

* **Copying `workload.status.clusterName` on scale up**

  * Fixed the race by moving the copy from workload creation to the scheduler replacement point, immediately before admitting the new slice.
  * Clarified the purpose, consistent placement and skipping nomination.

* **MultiKueue workload controller handling for `Finished` and `WithoutQuotaReservation` with ElasticJobs**

  * Described why behavior differs for ElasticJobs.
  * Reduced duplication and refactored the flow to be more intuitive.

ichekrygin commented:
/retest

ichekrygin commented:
/retest

ichekrygin commented:
/test pull-kueue-test-e2e-kueueviz-main

tenzen-y commented:
/test pull-kueue-test-e2e-main-1-31

ichekrygin commented:
/test pull-kueue-test-e2e-kueueviz-main
/test pull-kueue-test-e2e-main-1-32
/test pull-kueue-test-e2e-main-1-33
/test pull-kueue-test-e2e-main-1-34

tenzen-y commented on pkg/controller/admissionchecks/multikueue/workload.go:
```diff
@@ -95,6 +96,13 @@ func (g *wlGroup) IsFinished() bool {
 	return apimeta.IsStatusConditionTrue(g.local.Status.Conditions, kueue.WorkloadFinished)
 }
 
+// IsElasticWorkload returns true if the workload is considered elastic,
+// meaning the ElasticJobsViaWorkloadSlices feature is enabled and the
+// workload has the corresponding annotation set.
+func (g *wlGroup) IsElasticWorkload() bool {
+	return workloadslicing.IsElasticWorkload(g.local)
+}
+
```
```suggestion
```

Do we still need? In each place, couldn't we just call `workloadslicing.IsElasticWorkload(g.local)`?

ichekrygin replied:
I think this helper is consistent with other group functions (e.g., `IsFinished()`), where we don’t need to explicitly unpack group attributes.

That said, I can remove it if it’s a blocker.

mimowo replied:
Im ok either way. I like consistency

ichekrygin commented:
@tenzen-y, thank you for the comment (not sure why I couldn’t quote or reply directly, so responding here in a separate thread).

> The status field is the actual state, and spec is the desired state.
> However, if we assign any clusterName to the status field w/o QuotaReservation, the .status.clusterName represents the desired state since the .status.clusterName is no longer actual cluster name reserved by the cluster.

For a new workload slice, the **desired cluster name** is carried forward from the old workload slice’s `.status.clusterName`.
To ensure consistent placement of all scaled-up workloads, we want them to continue using the original placement (i.e., the same cluster name) and to skip cluster nomination altogether. We achieve this by setting `status.clusterName` on the new slice.

Does this explanation make sense?

tenzen-y commented:
> @tenzen-y, thank you for the comment (not sure why I couldn’t quote or reply directly, so responding here in a separate thread).
> 
> > The status field is the actual state, and spec is the desired state.
> > However, if we assign any clusterName to the status field w/o QuotaReservation, the .status.clusterName represents the desired state since the .status.clusterName is no longer actual cluster name reserved by the cluster.
> 
> For a new workload slice, the **desired cluster name** is carried forward from the old workload slice’s `.status.clusterName`. To ensure consistent placement of all scaled-up workloads, we want them to continue using the original placement (i.e., the same cluster name) and to skip cluster nomination altogether. We achieve this by setting `status.clusterName` on the new slice.
> 
> Does this explanation make sense?

My point is `.status.clusterName` must be the actual reserved cluster place. IIUC, in the current implementations, the WorkloadSlice's `.status.clusterName` is just desired since the clusterName is assigned before QuotaReservation in the worker cluster name?

I might be missing something. If my understanding is incorrect, please let me know.

mimowo commented:
Im also thinking that going via NominatedClusterNames can cause race conditions if the external dispatcher updates the value before it is promoted to status.clusterName. Maybe we could precent updating status.nominatedClusterNames for prebuilt workloads to prevent that phenomenon, but it seems ad hoc. Requiring all external Dispatchers to avoid mutating prebuilt workloads is not preferred if can be avoided.

mimowo commented on pkg/controller/admissionchecks/multikueue/workload.go:
```diff
@@ -442,22 +469,30 @@ func (w *wlReconciler) nominateAndSynchronizeWorkers(ctx context.Context, group
+	// supporting preferred or required placement constraints.
+	if clusterName := workload.ClusterName(group.local); group.IsElasticWorkload() && clusterName != "" {
+		nominatedWorkers = []string{clusterName}
+	} else {
+		if w.dispatcherName == config.MultiKueueDispatcherModeAllAtOnce {
```
I would prefer else-if or "switch" just to avoid indentation. On the first review pass seeing the last diff I was tricked to think the dispatchers logic is changed. Also, saving the indentation level is a win for readability on its own.

mimowo commented:
> I might be missing part of your point here. Do you feel we’re aligned with the PR change, or do you still see a concern? If it’s the latter, could you help point me to the specific area and what you’d recommend?

I think we are aligned to the part about skipping going via the nomination phase (status.nominatedClusterNames). My comment was actually describing there could be race conditions if we were going via the nomination phase (as suggested by @tenzen-y in https://github.com/kubernetes-sigs/kueue/pull/6445#discussion_r2382748870).

Even if we skip the nomination phase a race condition exists where the external dispatcher could nominate a cluster name and scheduler could admit before we propagate the .status.clusterName. As suggested in https://github.com/kubernetes-sigs/kueue/pull/6445#discussion_r2383033945 one idea of solving the issue is to prevent setting `status.nominatedClusterName` entirely for elastic prebuilt workloads.
